### DISCUSSION TITLE
URL pattern for addressing documents.

### DISCUSSION THREAD

**Discussion by HMarzban** [docs-plus/docs.plus#48] [Ideas] [answered]
## URL/Uniform resource locator

What I propose is this pattern:

`www.docs.plus/username/document_name`

_(Or we can replace `username` with the unique name and then name it `namespace`)_

This pattern is a way to create a more personalized and organized URL structure for addressing documents.

**Key features:**
1. **Personalization**: By including the user's name in the URL, it becomes more personal, and also it could be helpful to identify the user's documents.
2. **Organization**: Grouping documents by a user can make it easier to manage and organize the documents on the website.
3. **Access control**: By associating documents with specific users, it can be used to control access to the documents. Only the user or authorized users can access the document.
4. **Easy to remember**: If the document's URL includes the user's name, it may be easier for the user to remember and find the document again in the future.
5. **User-friendly**: It can make the URLs easier to read and understand.

However, there are also some potential downsides to this pattern. For example, if a user changes their name, the URL will no longer be accurate, and a redirect must be implemented to avoid broken links. Additionally, this pattern may not work well if the website has a large number of users or documents, as it could lead to a very deep directory structure.

Overall, this pattern is one way to address a document, and it could be helpful when the website is designed for personal usage or collaboration.

Ultimately, the best solution will depend on our specific needs and the trade-offs we are willing to make between personalization and scalability.

## Filters
The other subject that we have to consider is the filters!

For instance, we have a document with the name `document_name,` which belongs to the username `adem_po`; the URL pattern would be like this:
`www.docs.plus/adem_po/document_name`
In this document, I need to filter contents based on the Heading hierarchy with these words:

- democracy
- 2020
- china
- alternative  

We can show the filters (append filter parameters to the URL) in two ways:

1. **Query parameters:**
  1.1 `www.docs.plus/adem_po/document_name?filter=democracy`
  1.2 `www.docs.plus/adem_po/document_name?filter=democracy,2020,china,alternative`
2. **Subdirectory:**
  2.1 `www.docs.plus/adem_po/document_name/democracy`
  2.2 2.2 `www.docs.plus/adem_po/document_name/democracy/2020/china/alternative`

### Alternative, Use a hash fragment in the URL
This method allows us to include a filter in the URL after the "#" symbol, For example:

`www.docs.plus/username/docuemnt_name#filter=democracy,2020,china,alternative`

**Comment by edsaperia** [comment]
How about:

Default open pads can be at
`docs.plus/document_name`

Private pads at
`username.docs.plus/document_name`

Filters:
`username.docs.plus/document_name/filter1/filter2`

And to link to individual headings, something like:
`username.docs.plus/document_name#heading-ID-heading-text`

**Reply by HMarzban** [reply, marked as answer]
It's good to have the `username` as a subdomain, but it makes it difficult to maintain the production.
For now, we'll choose the `www.docs.plus/username/filters`. 
Then, we can consider this subdomain feature for the upcoming version.

Also, to link to individual headings, what you suggest is reasonable for me. Let's stick to this one.
